File: 3403.go
package main

func answerString(word string, numFriends int) string {
	if numFriends == 1 {
		return word
	}
	
	largestChar := 'a'
	ans := ""

	for _, c := range(word) {
		if c > largestChar {
			largestChar = c
		}
	}

	n := len(word)

	if numFriends == n {
		return string(largestChar)
	}

	maxLen := n - numFriends + 1

	for i := 0; i < n; i++ {
		if word[i] == byte(largestChar) {
            end := i + maxLen
            if end > n {
                end = n
            }
			
            possible := word[i:end]
            if possible > ans {
                ans = possible
            }
        }
	}
	return ans
}

func main() {
	answerString("aann", 2)
}

File: 3403.py
class Solution(object):
    def answerString(self, word, numFriends):
        """
        :type word: str
        :type numFriends: int
        :rtype: str
        """
        if numFriends == 1:
            return word
        
        largestChar = max(word)
        
        if numFriends == len(word):
            return largestChar
        
        ans = ""
        # l = 0
        n = len(word)
        maxLen = n - numFriends + 1
        
        for i, c in enumerate(word):
            if c == largestChar:
                possible = word[i:i+maxLen]
                
                ans = max(ans, possible)

        return ans
        
        # for r in range(n):
        #     ans = max(ans, word[l:r+1])
        #     # print(word[l:r+1])
            
        #     if len(word[l:r+1]) >= n - numFriends + 1:
        #         # while len(word[l:r+1]) > 1 and l <= r + 1:
        #         temp = word[l:r+1]
        #         ans = max(ans, word[l:r+1])
        #         l += 1
                
        #         left = 0
        #         while len(temp) > 1 and left <= len(temp):
        #             ans = max(ans, temp[left:len(temp)])
        #             # print(temp[left:len(temp)])
        #             left += 1

obj = Solution()
print(obj.answerString("epbbppl", 2))